Add color_write_enable sample

Krzysztof-Dmitruk-Mobica opened the pull request:
## Description

This PR adds a sample demonstrating the use of the VK_EXT_color_write_enable extension.

## Notes
Tested on:

Ubuntu 22.04
Intel i7-11800H
Intel TigerLake-H GT1
32 GB RAM

Ubuntu 22.04
Intel i7-11800H
Nvidia T600 Mobile
32 GB RAM

Windows 11
AMD Ryzen 2600
Nvidia GTX 1070
16 GB RAM

According to [vulkan.gpuinfo.org](https://vulkan.gpuinfo.org/) there is no macOS device on which VK_EXT_color_write_enable is supported.

~~There is a validation layer error described in the README.md.~~

## General Checklist:

Please ensure the following points are checked:

- [X] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [X] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [X] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [X] I have commented any added functions (in line with Doxygen)
- [X] I have commented any code that could be hard to understand
- [X] My changes do not add any new compiler warnings
- [X] My changes do not add any new validation layer errors or warnings
- [X] I have used existing framework/helper functions where possible
- [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [X] My changes do not add any regressions
- [X] I have tested every sample to ensure everything runs correctly
- [X] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [X] I have tested the sample on at least one compliant Vulkan implementation
- [X] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [X] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [X] Any dependent assets have been merged and published in downstream modules
- [X] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

Karina-Murawko-Wisniewska-Mobica commented on samples/extensions/color_write_enable/README.md:
```diff
@@ -0,0 +1,34 @@
+Adding multiple color blend attachments requires the addition of the corresponding color attachments in the render subpass. Each additional render subpass attachment requires an  extra framebuffer. These framebuffers are filled with data separately in the fragment shader.
+
+## The sample
+Thes ample shows how to setup an application to work with this extension:
```
```suggestion
The sample shows how to setup an application to work with this extension:
```

Krzysztof-Dmitruk-Mobica replied:
Done

Karina-Murawko-Wisniewska-Mobica commented on samples/extensions/color_write_enable/color_write_enable.cpp:
```diff
@@ -0,0 +1,331 @@
+
+void ColorWriteEnable::request_gpu_features(vkb::PhysicalDevice &gpu)
+{
+	auto &feature                        = gpu.request_extension_features<VkPhysicalDeviceVulkan12Features>(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_COLOR_WRITE_ENABLE_FEATURES_EXT);
```
Why VkPhysicalDeviceVulkan12Features is used instead of VkPhysicalDeviceColorWriteEnableFeaturesEXT? Also variable feature is not used later.

Krzysztof-Dmitruk-Mobica replied:
Done

Karina-Murawko-Wisniewska-Mobica commented on samples/extensions/color_write_enable/color_write_enable.cpp:
```diff
@@ -0,0 +1,331 @@
+void ColorWriteEnable::request_gpu_features(vkb::PhysicalDevice &gpu)
+{
+	auto &feature                        = gpu.request_extension_features<VkPhysicalDeviceVulkan12Features>(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_COLOR_WRITE_ENABLE_FEATURES_EXT);
+	gpu.get_mutable_requested_features() = gpu.get_features();
```
Is it necessary to enable all features?

Krzysztof-Dmitruk-Mobica replied:
It is not. In fact no features besides ColorWriteEnable are required.

Karina-Murawko-Wisniewska-Mobica commented on samples/extensions/color_write_enable/color_write_enable.cpp:
```diff
@@ -0,0 +1,331 @@
+				vkDestroyFramebuffer(device->get_handle(), framebuffers[i], nullptr);
+			}
+		}
+	}
```
It could be simplified, if framebuffers.size() is "0" the for loop will not be executed 

```suggestion
	for (uint32_t i = 0; i < framebuffers.size(); ++i)
	{
		if (framebuffers[i] != VK_NULL_HANDLE)
		{
			vkDestroyFramebuffer(device->get_handle(), framebuffers[i], nullptr);
		}
	}
```

Krzysztof-Dmitruk-Mobica replied:
Done

Karina-Murawko-Wisniewska-Mobica reviewed (commented):
Add sample in samples/CMakeLists.txt file.

gary-sweet reviewed (changes requested):
All I see when I run this sample is a triangle. That doesn't seem right given that there appears to be some UI component to it in the code.

![image](https://user-images.githubusercontent.com/20819159/233366213-fa569890-a10b-4b9c-8682-d5f65fb0fbce.png)

SaschaWillems reviewed (changes requested):
It works fine on my nvidia RTX 2060, but I do get validation errors which may hint at the problem Gary is seeing:

[error] [framework\core\instance.cpp:42] 96707441 - VUID-VkGraphicsPipelineCreateInfo-renderPass-06042: Validation Error: [ VUID-VkGraphicsPipelineCreateInfo-renderPass-06042 ] Object 0: handle = 0x24fe48f59f0, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x5c3a371 | vkCreateGraphicsPipelines() pCreateInfo[0]: VkRenderPass 0xab64de0000000020[] subpass 0 has colorAttachmentCount of 3 which doesn't match the pColorBlendState->attachmentCount of 1. The Vulkan spec states: If renderPass is not VK_NULL_HANDLE, and the pipeline is being created with fragment output interface state, and the subpass uses color attachments, the attachmentCount member of pColorBlendState must be equal to the colorAttachmentCount used to create subpass (https://vulkan.lunarg.com/doc/view/1.3.243.0/windows/1.3-extensions/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-renderPass-06042)

and

[error] [framework\core\instance.cpp:42] -884533293 - VUID-VkPipelineColorBlendStateCreateInfo-pAttachments-00605: Validation Error: [ VUID-VkPipelineColorBlendStateCreateInfo-pAttachments-00605 ] Object 0: handle = 0x24fe48f59f0, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xcb4717d3 | vkCreateGraphicsPipelines() pCreateInfo[0]: If independent blend feature not enabled, all elements of pAttachments must be identical. The Vulkan spec states: If the independentBlend feature is not enabled, all elements of pAttachments must be identical (https://vulkan.lunarg.com/doc/view/1.3.243.0/windows/1.3-extensions/vkspec.html#VUID-VkPipelineColorBlendStateCreateInfo-pAttachments-00605)

Krzysztof-Dmitruk-Mobica commented:
It saddened me that the sample does not work for @gary-sweet. It looks like only the last attachment to which channel B is assigned is displayed. The UI is displayed on the first attachment, which is probably why you can't see it. At least from what I can see from your screenshot, blending works as intended. Please specify your hardware configuration. I'll see if I can reproduce something close to it.

gary-sweet commented:
I get the same result with the latest updates. I also get the same validation error that Sascha posted above.

gary-sweet commented:
> Please specify your hardware configuration. I'll see if I can reproduce something close to it.

My hardware config can be seen here: https://vulkan.gpuinfo.org/displayreport.php?id=19987

gary-sweet commented:
The independent blend validation error can be fixed by requesting the independent blend feature.

The missing menu problem is very likely related to the other validation error. You are drawing the UI using the same render pass setup that you are using for the main drawing. This has 3 attachments. The UI drawing is presumably only expecting to render to a single attachment. I suspect that you'll need to separate the UI drawing into another render pass.

asuessenbach commented on samples/extensions/color_write_enable/README.md:
```diff
@@ -0,0 +1,53 @@
+
+# Color write enable
+## Overview
+This sample demonstrates how to use `VK_EXT_color_write_enable` extension. This extension allows toggle the output color attachments using a pipeline dynamic state. It allows the program to prepare an additional framebuffer populated with the data from a defined color blend attachment which can be blended dynamically to the final scene.
```
how to use **the** `VK_EXT_color_write_enable`
allows **to** toggle

Krzysztof-Dmitruk-Mobica replied:
Fixed, thank you.

asuessenbach commented on samples/extensions/color_write_enable/README.md:
```diff
@@ -0,0 +1,53 @@
+The final results are comparable to those obtained with `vkCmdSetColorWriteMaskEXT`, but it does not require the GPU driver to support `VK_EXT_extended_dynamic_state3`.
+
+## How to use in Vulkan
+Program needs to enable extension `VK_EXT_COLOR_WRITE_ENABLE_EXTENSION_NAME` and request feature `VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_COLOR_WRITE_ENABLE_FEATURES_EXT`. 
```
To use this feature, the device extension `VK_EXT_COLOR_WRITE_ENABLE_EXTENSION_NAME` has to be enabled. Support of this feature can be queried by extending the struct `VkPhysicalDeviceFeatures2` in the `vkGetPhysicalDeviceFeatures2` call by a `VkPhysicalDeviceColorWriteEnableFeaturesEXT` struct.

Krzysztof-Dmitruk-Mobica replied:
Fixed, thank you.

asuessenbach commented on samples/extensions/color_write_enable/README.md:
```diff
@@ -0,0 +1,53 @@
+
+## How to use in Vulkan
+Program needs to enable extension `VK_EXT_COLOR_WRITE_ENABLE_EXTENSION_NAME` and request feature `VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_COLOR_WRITE_ENABLE_FEATURES_EXT`. 
+`VkPipelineColorWriteCreateInfoEXT` contains an array of Boolean values that serve as toggles for the corresponding `VkPipelineColorBlendAttachmentState`. This array can be overwritten dynamically with the `vkCmdSetColorWriteEnableEXT function`. 
```
`vkCmdSetColorWriteEnableEXT function` -> `vkCmdSetColorWriteEnableEXT` function

Krzysztof-Dmitruk-Mobica replied:
Fixed, thank you.

asuessenbach commented on samples/extensions/color_write_enable/color_write_enable.h:
```diff
@@ -0,0 +1,56 @@
+class ColorWriteEnable : public ApiVulkanSample
+{
+  public:
+	bool  r_bit_enabled      = true;
```
Shouldn't these values be private?

Krzysztof-Dmitruk-Mobica replied:
They definitely should be. I have fixed them.

Krzysztof-Dmitruk-Mobica committed:
Add color_write_enable sample

Krzysztof-Dmitruk-Mobica committed:
Refactor sample to use subpasses

- Add subpass with dynamic color write toggle
- Add composition + UI subpass
- Minor changes to the framework to allow change of the UI subpass

Krzysztof-Dmitruk-Mobica commented:
I did a major refactor of the sample. The render is now split into two subpasses. First subpass creates triangles in separate color attachments, and the second subpass combines them. The subpass code is mainly based on the subpass sample from [Sascha Willems' repo](https://github.com/SaschaWillems/Vulkan/blob/master/examples/subpasses/subpasses.cpp). The validation layer error has been fixed, so hopefully the error @gary-sweet experienced is fixed aswell.

gary-sweet reviewed (dismissed):
Looks good to me now. Thanks

cforfang commented:
HI @Krzysztof-Dmitruk-Mobica, after changing to use sub-passes, I noticed one bug is the composition shader is still using texture-sampling to read the input data -- this needs to use input-attachments instead (and then also need to change the subpass-dependency to be COLOR_ATTACHMENT_WRITE -> INPUT_ATTACHMENT_READ). 

EDIT:
And to mention it to avoid confusion, the use of VK_ACCESS_SHADER_READ_BIT in the SaschaWillems-sample is incorrect -- one needs to use COLOR_ATTACHMENT_WRITE  -> INPUT_ATTACHMENT_READ to get the legal and correct behavior on tilers.

Krzysztof-Dmitruk-Mobica committed:
Fix composition shader to use subpassInput type

Krzysztof-Dmitruk-Mobica commented:
Hi @cforfang, thanks for showing me the problem. Would you please look into the latest commit to see if I got it right?

cforfang commented:
> Hi @cforfang, thanks for showing me the problem. Would you please look into the latest commit to see if I got it right?

With the changes it looks good to me 👍 (I didn't try to run or check validation to be clear, but those are the exact changes which comes to mind 🙂)

asuessenbach commented on framework/api_vulkan_sample.h:
```diff
@@ -340,6 +340,11 @@ class ApiVulkanSample : public vkb::VulkanSample
+	/**
+	 * @brief Initializes the UI. Can be overridden to customize the way it is displayed.
+	 */
+	virtual void prepare_gui(vkb::Platform &platform);
```
Would you mind to add that function to hpp_api_vulkan_sample as well?

Krzysztof-Dmitruk-Mobica replied:
Sure, done.

asuessenbach commented on framework/gui.h:
```diff
@@ -349,6 +349,9 @@ class Gui
 	bool is_debug_view_active() const;
 
+  protected:
+	uint32_t subpass = 0;
```
Maybe make `subpass` private, and add a getter and a setter function?
In any case, would you mind to add that to hpp_gui as well?

Krzysztof-Dmitruk-Mobica replied:
Done.

asuessenbach commented on samples/extensions/color_write_enable/README.md:
```diff
@@ -0,0 +1,52 @@
+Support of this feature can be queried by extending the struct VkPhysicalDeviceFeatures2 in the vkGetPhysicalDeviceFeatures2 call by a VkPhysicalDeviceColorWriteEnableFeaturesEXT struct.
+`VkPipelineColorWriteCreateInfoEXT` contains an array of Boolean values that serve as toggles for the corresponding `VkPipelineColorBlendAttachmentState`. This array can be overwritten dynamically with the `vkCmdSetColorWriteEnableEXT` function. 
+
+Two subpasses are prerformed in the sample. In the first subpass, three attachments are used. Each attachment has only one color component bit enabled - R, G and B. Triangle is drawn on each of them separately.
```
typo: p**r**erformed
Missing A: **A** triangle is drawn...

Krzysztof-Dmitruk-Mobica replied:
Thanks, done aswell.

asuessenbach commented on samples/extensions/color_write_enable/color_write_enable.cpp:
```diff
@@ -0,0 +1,592 @@
+		attachments[0] = swapchain_buffers[i].view;
+		attachments[1] = this->attachments.red.view;
+		attachments[2] = this->attachments.green.view;
+		attachments[3] = this->attachments.blue.view;
```
Maybe move setting of `attachments[1...3]` out of the loop?

Krzysztof-Dmitruk-Mobica replied:
Done.

asuessenbach commented on samples/extensions/color_write_enable/color_write_enable.h:
```diff
@@ -0,0 +1,111 @@
+ * @brief Color generation toggle using VK_EXT_color_write_enable
+ */
+
+class GuiInSubpass1 : public vkb::Gui
```
There's no need to introduce a class `GuiInSubpass1`, derived from `vkb::Gui`, just to set some `subpass` value. As that value has become a member of `vkb::Gui`, you can just set it in `ColorWriteEnable::prepare_gui` (by some to be introduced setter function).

Krzysztof-Dmitruk-Mobica replied:
Done.

asuessenbach commented on shaders/color_write_enable/composition.frag:
```diff
@@ -0,0 +1,35 @@
+layout (input_attachment_index = 1, binding = 1) uniform subpassInput in_color_g;
+layout (input_attachment_index = 2, binding = 2) uniform subpassInput in_color_b;
+
+layout (location = 0) in vec2 inUV;
```
`inUV` isn't used here.

Krzysztof-Dmitruk-Mobica replied:
I forgot about it, done.

asuessenbach commented on shaders/color_write_enable/composition.vert:
```diff
@@ -0,0 +1,32 @@
+
+precision mediump float;
+
+layout (location = 0) out vec2 outUV;
```
This output isn't used in composition.frag.

Krzysztof-Dmitruk-Mobica replied:
Done.

asuessenbach commented on shaders/color_write_enable/triangle_separate_channels.frag:
```diff
@@ -0,0 +1,32 @@
+
+void main()
+{
+       out_color_r = vec4(in_color, 1.0);
```
Maybe a short comment that each of the three outputs is specified to use just one component (via the `blend_attachment` in `ColorWriteEnables::prepare_pipelines`)

Krzysztof-Dmitruk-Mobica replied:
Done.

asuessenbach commented on shaders/color_write_enable/composition.frag:
```diff
@@ -0,0 +1,35 @@
+ * limitations under the License.
+ */
+
+precision mediump float;
```
Maybe a short comment why medium precision is ok here.

Krzysztof-Dmitruk-Mobica replied:
I chose to remove the precision setting as it is unimportant for this sample. In order to do so, I had to upgrade the version in the directive.

Krzysztof-Dmitruk-Mobica committed:
Fix multiple problems pointed by asuessenbach
